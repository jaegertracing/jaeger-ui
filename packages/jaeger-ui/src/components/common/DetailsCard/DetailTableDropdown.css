/*
Copyright (c) 2019 Uber Technologies, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.DetailTableDropdown--Footer {
  background: white;
  border: 1px solid #ccc;
  display: flex;
  justify-content: space-between;
  padding: 0.3em;
}

.DetailTableDropdown--Footer--CancelConfirm {
  display: flex;
}

.DetailTableDropdown--Btn {
  align-items: center;
  border-radius: 0;
  color: #ddd;
  display: flex;
  padding: 0.3em 0.7em 0.3em 0.6em;
}

.DetailTableDropdown--Btn ~ .DetailTableDropdown--Btn {
  margin-left: 0.3em;
}

.DetailTableDropdown--Btn > *:not(:first-child) {
  margin-left: 0.3em;
}

.DetailTableDropdown--Btn.Apply {
  background: #4c21ce;
}

.DetailTableDropdown--Btn.Cancel {
  background: #007272;
}

.DetailTableDropdown--Btn.Clear {
  background: #ff2626;
}

.DetailTableDropdown--Tooltip {
  max-width: unset;
  white-space: nowrap;
}

.DetailTableDropdown--Tooltip--Body {
  display: flex;
  flex-direction: column;
}
