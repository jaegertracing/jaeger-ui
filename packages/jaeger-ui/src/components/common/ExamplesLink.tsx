// Copyright (c) 2020 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';

import { getUrl } from '../SearchTracePage/url';
import NewWindowIcon from './NewWindowIcon';

export type TExample = {
  spanIDs?: string[];
  traceID: string;
};

type TProps = {
  examples?: TExample[];
  includeText?: boolean;
};

type TExampleWithSpans = {
  traceID: string;
  spanIDs: string[];
};

function hasSpans(example: TExample | TExampleWithSpans): example is TExampleWithSpans {
  return Boolean(example.spanIDs && example.spanIDs.length);
}

function getGetUrlArg(examples: TExample[]): { spanLinks: Record<string, string>; traceID: string[] } {
  const spanLinks: Record<string, string> = {};
  const traceID: string[] = [];

  examples.forEach((example: TExample) => {
    if (hasSpans(example)) spanLinks[example.traceID] = example.spanIDs.join(' ');
    else traceID.push(example.traceID);
  });

  return {
    spanLinks,
    traceID,
  };
}

export default class ExamplesLink extends React.PureComponent<TProps> {
  render() {
    const { examples, includeText } = this.props;

    if (!examples || !examples.length) return null;

    return (
      <a href={getUrl(getGetUrlArg(examples))} target="_blank" rel="noopener noreferrer">
        {includeText && 'Examples '}
        <NewWindowIcon />
      </a>
    );
  }
}
