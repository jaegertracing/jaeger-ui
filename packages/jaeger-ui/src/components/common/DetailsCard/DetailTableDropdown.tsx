// Copyright (c) 2020 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import { Button, Tooltip } from 'antd';
import FaCheck from 'react-icons/lib/fa/check.js';
import FaTrash from 'react-icons/lib/fa/trash.js';
import TiCancel from 'react-icons/lib/ti/cancel.js';

import FilteredList from '../FilteredList';

import { TFilterDropdownProps } from './types';

import './DetailTableDropdown.css';

type TProps = TFilterDropdownProps & {
  options: Set<string>;
};

export default class DetailTableDropdown extends React.PureComponent<TProps> {
  cancelled = false;
  selected: React.Key[] = [];

  componentDidUpdate(prevProps: TProps) {
    const { confirm, selectedKeys } = this.props;

    // If the entries in selectedKeys is unchanged, the dropdown has opened or closed.
    // Record the selectedKeys at this time for future cancellations.
    if (selectedKeys.length === prevProps.selectedKeys.length) {
      const prevKeys = new Set(prevProps.selectedKeys);
      if (selectedKeys.every(key => prevKeys.has(key))) {
        this.selected = selectedKeys;
      }
    }

    // Unfortunately antd requires setSelectedKeys and confirm to be called in different cycles.
    if (this.cancelled) {
      this.cancelled = false;
      confirm();
    }
  }

  cancel = () => {
    // Unfortunately antd requires setSelectedKeys and confirm to be called in different cycles.
    this.cancelled = true;
    this.props.setSelectedKeys(this.selected);
  };

  render() {
    const { clearFilters = () => {}, confirm, options, selectedKeys, setSelectedKeys } = this.props;

    const value = new Set<string>();
    selectedKeys.forEach(selected => {
      if (typeof selected === 'string') value.add(selected);
    });

    return (
      <div>
        <FilteredList
          addValues={(values: string[]) => {
            setSelectedKeys([...selectedKeys, ...values]);
          }}
          multi
          options={Array.from(options)}
          removeValues={(values: string[]) => {
            const remove = new Set<React.Key>(values);
            setSelectedKeys(selectedKeys.filter(key => !remove.has(key)));
          }}
          setValue={(v: string) => {
            setSelectedKeys([v]);
          }}
          value={value}
        />
        <div className="DetailTableDropdown--Footer">
          <Tooltip overlayClassName="DetailTableDropdown--Tooltip" title="Remove filter from this column">
            <Button className="DetailTableDropdown--Btn Clear" onClick={clearFilters}>
              <FaTrash size={18} />
              Clear Filter
            </Button>
          </Tooltip>
          <div className="DetailTableDropdown--Footer--CancelConfirm">
            <Tooltip
              overlayClassName="DetailTableDropdown--Tooltip"
              title="Cancel changes to this column's filter"
            >
              <Button className="DetailTableDropdown--Btn Cancel" onClick={this.cancel}>
                <TiCancel size={20} />
                Cancel
              </Button>
            </Tooltip>
            <Tooltip
              overlayClassName="DetailTableDropdown--Tooltip"
              title={
                <div className="DetailTableDropdown--Tooltip--Body">
                  <span>Apply changes to this column{"'"}s filter</span>
                  <span>Same effect as clicking outside the dropdown</span>
                </div>
              }
            >
              <Button className="DetailTableDropdown--Btn Apply" onClick={confirm}>
                <FaCheck size={18} />
                Apply
              </Button>
            </Tooltip>
          </div>
        </div>
      </div>
    );
  }
}
