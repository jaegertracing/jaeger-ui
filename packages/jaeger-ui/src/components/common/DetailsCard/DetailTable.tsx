// Copyright (c) 2020 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import { Icon, Table } from 'antd';
import FaFilter from 'react-icons/lib/fa/filter.js';
import _isEmpty from 'lodash/isEmpty';

import ExamplesLink, { TExample } from '../ExamplesLink';
import DetailTableDropdown from './DetailTableDropdown';

import { TColumnDef, TColumnDefs, TFilterDropdownProps, TRow, TStyledValue } from './types';

// exported for tests
export const _makeFilterDropdown = (dataIndex: string, options: Set<string>) => (
  props: TFilterDropdownProps
) => {
  return <DetailTableDropdown {...props} key={dataIndex} options={options} />;
};

// exported for tests
export const _onCell = (dataIndex: string) => (row: TRow) => {
  const cellData = row[dataIndex];
  if (!cellData || typeof cellData !== 'object' || Array.isArray(cellData)) return null;
  const { styling } = cellData;
  if (_isEmpty(styling)) return null;
  return {
    style: styling,
  };
};

// exported for tests
export const _onFilter = (dataIndex: string) => (value: string, row: TRow) => {
  const data = row[dataIndex];
  if (typeof data === 'object' && !Array.isArray(data) && typeof data.value === 'string') {
    return data.value === value;
  }
  return data === value;
};

// exported for tests
export const _renderCell = (cellData: undefined | string | TStyledValue) => {
  if (!cellData || typeof cellData !== 'object') return cellData;
  if (Array.isArray(cellData)) return <ExamplesLink examples={cellData} />;
  if (!cellData.linkTo) return cellData.value;
  return (
    <a href={cellData.linkTo} target="_blank" rel="noopener noreferrer">
      {cellData.value}
    </a>
  );
};

// exported for tests
export const _sort = (dataIndex: string) => (a: TRow, b: TRow) => {
  const aData = a[dataIndex];
  let aValue;
  if (Array.isArray(aData)) aValue = aData.length;
  else if (typeof aData === 'object' && typeof aData.value === 'string') aValue = aData.value;
  else aValue = aData;

  const bData = b[dataIndex];
  let bValue;
  if (Array.isArray(bData)) bValue = bData.length;
  else if (typeof bData === 'object' && typeof bData.value === 'string') bValue = bData.value;
  else bValue = bData;

  if (aValue < bValue) return -1;
  return bValue < aValue ? 1 : 0;
};

// exported for tests
export const _makeColumns = ({ defs, rows }: { defs: TColumnDefs; rows: TRow[] }) =>
  defs.map((def: TColumnDef | string) => {
    let dataIndex: string;
    let key: string;
    let sortable: boolean = true;
    let style: React.CSSProperties | undefined;
    let title: string;
    if (typeof def === 'string') {
      // eslint-disable-next-line no-multi-assign
      key = title = dataIndex = def;
    } else {
      // eslint-disable-next-line no-multi-assign
      key = title = dataIndex = def.key;
      if (def.label) title = def.label;
      if (def.styling) style = def.styling;
      if (def.preventSort) sortable = false;
    }

    const options = new Set<string>();
    rows.forEach(row => {
      const value = row[dataIndex];
      if (typeof value === 'string' && value) options.add(value);
      else if (typeof value === 'object' && !Array.isArray(value) && typeof value.value === 'string') {
        options.add(value.value);
      }
    });

    return {
      dataIndex,
      key,
      title,
      filterDropdown: Boolean(options.size) && _makeFilterDropdown(dataIndex, options),
      filterIcon: (filtered: boolean) => {
        if (filtered) return <FaFilter />;
        return <Icon type="filter" />;
      },
      onCell: _onCell(dataIndex),
      onHeaderCell: () => ({
        style,
      }),
      onFilter: _onFilter(dataIndex),
      render: _renderCell,
      sorter: sortable && _sort(dataIndex),
    };
  });

// exported for tests
export const _rowKey = (row: TRow) =>
  JSON.stringify(row, function replacer(
    key: string,
    value: TRow | undefined | string | number | TStyledValue | TExample[]
  ) {
    function isRow(v: typeof value): v is TRow {
      return v === row;
    }
    if (isRow(value)) return value;
    if (Array.isArray(value)) return JSON.stringify(value);
    if (typeof value === 'object') {
      if (typeof value.value === 'string') return JSON.stringify(value);
      return value.value.key || 'Unknown';
    }
    return value;
  });

export default function DetailTable({
  columnDefs: _columnDefs,
  details,
}: {
  columnDefs?: TColumnDefs;
  details: TRow[];
}) {
  const columnDefs: TColumnDefs = _columnDefs ? _columnDefs.slice() : [];
  const knownColumns = new Set(
    columnDefs.map(keyOrObj => {
      if (typeof keyOrObj === 'string') return keyOrObj;
      return keyOrObj.key;
    })
  );
  details.forEach(row => {
    Object.keys(row).forEach((col: string) => {
      if (!knownColumns.has(col)) {
        knownColumns.add(col);
        columnDefs.push(col);
      }
    });
  });

  return (
    <Table
      key="table"
      size="middle"
      columns={_makeColumns({ defs: columnDefs, rows: details })}
      dataSource={details}
      pagination={false}
      rowKey={_rowKey}
    />
  );
}
