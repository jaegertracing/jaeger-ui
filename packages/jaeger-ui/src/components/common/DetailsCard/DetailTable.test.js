// Copyright (c) 2020 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import React from 'react';
import { shallow } from 'enzyme';
import { Icon } from 'antd';
import FaFilter from 'react-icons/lib/fa/filter.js';

import ExamplesLink from '../ExamplesLink';
import DetailTableDropdown from './DetailTableDropdown';
import DetailTable, {
  _makeColumns,
  _makeFilterDropdown,
  _onCell,
  _onFilter,
  _renderCell,
  _rowKey,
  _sort,
} from './DetailTable';

describe('DetailTable', () => {
  describe('render', () => {
    const col0 = {
      styling: {
        background: 'red',
        color: 'white',
      },
      key: 'col0',
    };
    const col1 = 'col1';
    const row0 = {
      [col0.key]: 'val0',
      [col1]: 'val1',
    };
    const row1 = {
      [col0.key]: 'val2',
      [col1]: 'val3',
    };

    it('renders given rows and columns', () => {
      expect(shallow(<DetailTable columnDefs={[col1, col0]} details={[row0, row1]} />)).toMatchSnapshot();
    });

    it('infers all columns', () => {
      expect(shallow(<DetailTable details={[row0]} />)).toMatchSnapshot();
    });

    it('infers missing columns', () => {
      expect(shallow(<DetailTable columnDefs={[col0]} details={[row0]} />)).toMatchSnapshot();
    });

    it('does not duplicate columns', () => {
      expect(shallow(<DetailTable columnDefs={[col1]} details={[row0, row1]} />)).toMatchSnapshot();
    });
  });

  describe('_rowKey', () => {
    const column = 'col';
    const examples = [{ spanIDs: ['id0', 'id1'], traceID: 'traceID' }];

    it('handles undefined', () => {
      const row = { [column]: undefined };
      expect(_rowKey(row)).toBe(JSON.stringify(row));
    });

    it('handles array', () => {
      const row = { [column]: examples };
      expect(_rowKey(row)).toBe(JSON.stringify({ [column]: JSON.stringify(examples) }));
    });

    it('handles object with string value', () => {
      const valueObject = { value: 'test-value' };
      const row = { [column]: valueObject };
      expect(_rowKey(row)).toBe(JSON.stringify({ [column]: JSON.stringify(valueObject) }));
    });

    it('handles object with React.Element value with key', () => {
      const key = 'test-key';
      const elem = <ExamplesLink key={key} examples={examples} />;
      const row = { [column]: { value: elem } };
      expect(_rowKey(row)).toBe(JSON.stringify({ [column]: key }));
    });

    it('handles object with React.Element value without key', () => {
      const elem = <ExamplesLink examples={examples} />;
      const row = { [column]: { value: elem } };
      expect(_rowKey(row)).toBe(JSON.stringify({ [column]: 'Unknown' }));
    });
  });

  describe('_makeColumns', () => {
    const stringColumn = 'stringCol';

    describe('static props', () => {
      const makeColumn = (def, rows = []) => _makeColumns({ defs: [def], rows })[0];

      it('renders string column', () => {
        expect(makeColumn(stringColumn)).toEqual({
          dataIndex: stringColumn,
          key: stringColumn,
          title: stringColumn,
          filterDropdown: false,
          filterIcon: expect.any(Function),
          onFilter: expect.any(Function),
          onCell: expect.any(Function),
          onHeaderCell: expect.any(Function),
          render: expect.any(Function),
          sorter: expect.any(Function),
        });
      });

      it('renders object column', () => {
        expect(makeColumn({ key: stringColumn })).toEqual({
          dataIndex: stringColumn,
          key: stringColumn,
          title: stringColumn,
          filterDropdown: false,
          filterIcon: expect.any(Function),
          onFilter: expect.any(Function),
          onCell: expect.any(Function),
          onHeaderCell: expect.any(Function),
          render: expect.any(Function),
          sorter: expect.any(Function),
        });
      });

      it('renders object column with label', () => {
        const label = `label, not-${stringColumn}`;
        expect(
          makeColumn({
            key: stringColumn,
            label,
          })
        ).toEqual({
          dataIndex: stringColumn,
          key: stringColumn,
          title: label,
          filterDropdown: false,
          filterIcon: expect.any(Function),
          onFilter: expect.any(Function),
          onCell: expect.any(Function),
          onHeaderCell: expect.any(Function),
          render: expect.any(Function),
          sorter: expect.any(Function),
        });
      });

      it('renders object column with styling', () => {
        const styling = {
          background: 'red',
          color: 'white',
        };
        expect(
          makeColumn({
            key: stringColumn,
            styling,
          }).onHeaderCell().style
        ).toBe(styling);
      });

      it('renders filter icon without filter set', () => {
        const icon = makeColumn(stringColumn).filterIcon();
        expect(icon.type).toBe(Icon);
        expect(icon.props.type).toBe('filter');
      });

      it('renders filter icon with filter set', () => {
        const icon = makeColumn(stringColumn).filterIcon(true);
        expect(icon.type).toBe(FaFilter);
      });

      it('renders filterable column if there are filterable values', () => {
        const filterableValues = ['foo', 'bar', { value: 'obj foo' }, { value: 'obj baz' }];
        const expected = new Set(filterableValues.map(v => v.value || v));
        const values = [
          ...filterableValues,
          <ExamplesLink examples={[]} />,
          <ExamplesLink key="fookey" examples={[]} />,
          undefined,
        ];
        const rows = values.map(value => ({
          [stringColumn]: value,
        }));
        const column = makeColumn(stringColumn, rows);
        const dropdown = column.filterDropdown();

        expect(dropdown.props.options).toEqual(expected);
        expect(dropdown.type).toBe(DetailTableDropdown);
      });

      it('renders object column without sort', () => {
        expect(
          makeColumn({
            key: stringColumn,
            preventSort: true,
          })
        ).toEqual({
          dataIndex: stringColumn,
          key: stringColumn,
          title: stringColumn,
          filterDropdown: false,
          filterIcon: expect.any(Function),
          onFilter: expect.any(Function),
          onCell: expect.any(Function),
          onHeaderCell: expect.any(Function),
          render: expect.any(Function),
          sorter: false,
        });
      });
    });

    describe('function props', () => {
      const makeTestFn = fn => (...vals) =>
        fn(stringColumn)(
          ...vals.map(v => ({
            [stringColumn]: v,
          }))
        );

      describe('_makeFilterDropdown', () => {
        it('returns DetailsTableDropdown with correct props', () => {
          const options = ['foo', 'bar'];
          const mockAntdDropdownProps = {
            foo: 'bar',
            bar: 'baz',
            baz: 'foo',
          };
          const filterDropdown = _makeFilterDropdown(stringColumn, options)(mockAntdDropdownProps);

          expect(filterDropdown.type).toBe(DetailTableDropdown);
          expect(filterDropdown.props).toEqual({
            ...mockAntdDropdownProps,
            options,
          });
          expect(filterDropdown.key).toBe(stringColumn);
        });
      });

      describe('_onCell', () => {
        const onCell = makeTestFn(_onCell);

        it('returns null for undefined', () => {
          expect(onCell(undefined)).toBe(null);
        });

        it('returns null for string', () => {
          expect(onCell('test-string')).toBe(null);
        });

        it('returns null for array', () => {
          expect(onCell([])).toBe(null);
        });

        it('returns null for unstyled object', () => {
          expect(onCell({})).toBe(null);
          expect(onCell({ styling: {} })).toBe(null);
        });

        it('returns styling for styled object', () => {
          const styling = {
            background: 'red',
            color: 'white',
          };
          expect(onCell({ styling }).style).toBe(styling);
        });
      });

      describe('_onFilter', () => {
        const onFilter = (filterValue, testValue) =>
          _onFilter(stringColumn)(filterValue, { [stringColumn]: testValue });
        const value = 'test-value';

        it('returns true if string value is filter', () => {
          expect(onFilter(value, value)).toBe(true);
        });

        it('returns true if object value is filter', () => {
          expect(onFilter(value, { value })).toBe(true);
        });

        it('returns false if string value is not filter', () => {
          expect(onFilter(value, `not-${value}`)).toBe(false);
        });

        it('returns false if object value is not filter', () => {
          expect(onFilter(value, { value: `not-${value}` })).toBe(false);
        });

        it('returns false for array value', () => {
          expect(onFilter(value, [value])).toBe(false);
        });

        it('returns false for undefined value', () => {
          expect(onFilter(value)).toBe(false);
        });
      });

      describe('_renderCell', () => {
        it('renders a string', () => {
          expect(_renderCell('a')).toBe('a');
        });

        it('handles undefined', () => {
          expect(_renderCell()).toBe(undefined);
        });

        it("renders an object's value", () => {
          const value = 'test-value';
          expect(_renderCell({ value })).toBe(value);
        });

        it('renders <ExamplesLink />', () => {
          const examples = [];
          const exampleLink = _renderCell(examples);
          expect(exampleLink.type).toBe(ExamplesLink);
          expect(exampleLink.props.examples).toBe(examples);
        });

        it('renders a regular link', () => {
          expect(_renderCell({ linkTo: 'test.link', value: 'test-value' })).toMatchSnapshot();
        });
      });

      describe('_sort ', () => {
        const sort = makeTestFn(_sort);

        it('sorts strings', () => {
          expect(sort('a', 'b')).toBe(-1);
          expect(sort('a', 'a')).toBe(0);
          expect(sort('b', 'a')).toBe(1);
        });

        it('sorts arrays by length', () => {
          expect(sort(new Array(2), new Array(3))).toBe(-1);
          expect(sort(new Array(2), new Array(2))).toBe(0);
          expect(sort(new Array(3), new Array(2))).toBe(1);
        });

        it('sorts objects with string values', () => {
          expect(sort({ value: 'a' }, { value: 'b' })).toBe(-1);
          expect(sort({ value: 'a' }, { value: 'a' })).toBe(0);
          expect(sort({ value: 'b' }, { value: 'a' })).toBe(1);
        });

        it('handles objects without string values', () => {
          expect(() => sort({}, {})).not.toThrow();
        });

        describe('mixed types', () => {
          it('sorts a string and an object', () => {
            expect(sort('a', { value: 'b' })).toBe(-1);
            expect(sort('a', { value: 'a' })).toBe(0);
            expect(sort('b', { value: 'a' })).toBe(1);
          });

          it('sorts an object and an array', () => {
            expect(sort({ value: 'a' }, new Array(3))).toBe(0);
            expect(sort({ value: 'a' }, new Array(2))).toBe(0);
            expect(sort({ value: 'b' }, new Array(2))).toBe(0);
          });

          it('sorts an array and a string', () => {
            expect(sort(new Array(2), 'a')).toBe(0);
            expect(sort(new Array(2), 'a')).toBe(0);
            expect(sort(new Array(3), 'b')).toBe(0);
          });
        });
      });
    });
  });
});
