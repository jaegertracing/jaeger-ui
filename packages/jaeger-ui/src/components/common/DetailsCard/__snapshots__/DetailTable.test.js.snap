// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailTable _makeColumns function props _renderCell renders a regular link 1`] = `
<a
  href="test.link"
  rel="noopener noreferrer"
  target="_blank"
>
  test-value
</a>
`;

exports[`DetailTable render does not duplicate columns 1`] = `
<ForwardRef(InternalTable)
  columns={
    Array [
      Object {
        "dataIndex": "col1",
        "filterDropdown": [Function],
        "filterIcon": [Function],
        "key": "col1",
        "onCell": [Function],
        "onFilter": [Function],
        "onHeaderCell": [Function],
        "render": [Function],
        "sorter": [Function],
        "title": "col1",
      },
      Object {
        "dataIndex": "col0",
        "filterDropdown": [Function],
        "filterIcon": [Function],
        "key": "col0",
        "onCell": [Function],
        "onFilter": [Function],
        "onHeaderCell": [Function],
        "render": [Function],
        "sorter": [Function],
        "title": "col0",
      },
    ]
  }
  dataSource={
    Array [
      Object {
        "col0": "val0",
        "col1": "val1",
      },
      Object {
        "col0": "val2",
        "col1": "val3",
      },
    ]
  }
  key="table"
  pagination={false}
  rowKey={[Function]}
  size="middle"
/>
`;

exports[`DetailTable render infers all columns 1`] = `
<ForwardRef(InternalTable)
  columns={
    Array [
      Object {
        "dataIndex": "col0",
        "filterDropdown": [Function],
        "filterIcon": [Function],
        "key": "col0",
        "onCell": [Function],
        "onFilter": [Function],
        "onHeaderCell": [Function],
        "render": [Function],
        "sorter": [Function],
        "title": "col0",
      },
      Object {
        "dataIndex": "col1",
        "filterDropdown": [Function],
        "filterIcon": [Function],
        "key": "col1",
        "onCell": [Function],
        "onFilter": [Function],
        "onHeaderCell": [Function],
        "render": [Function],
        "sorter": [Function],
        "title": "col1",
      },
    ]
  }
  dataSource={
    Array [
      Object {
        "col0": "val0",
        "col1": "val1",
      },
    ]
  }
  key="table"
  pagination={false}
  rowKey={[Function]}
  size="middle"
/>
`;

exports[`DetailTable render infers missing columns 1`] = `
<ForwardRef(InternalTable)
  columns={
    Array [
      Object {
        "dataIndex": "col0",
        "filterDropdown": [Function],
        "filterIcon": [Function],
        "key": "col0",
        "onCell": [Function],
        "onFilter": [Function],
        "onHeaderCell": [Function],
        "render": [Function],
        "sorter": [Function],
        "title": "col0",
      },
      Object {
        "dataIndex": "col1",
        "filterDropdown": [Function],
        "filterIcon": [Function],
        "key": "col1",
        "onCell": [Function],
        "onFilter": [Function],
        "onHeaderCell": [Function],
        "render": [Function],
        "sorter": [Function],
        "title": "col1",
      },
    ]
  }
  dataSource={
    Array [
      Object {
        "col0": "val0",
        "col1": "val1",
      },
    ]
  }
  key="table"
  pagination={false}
  rowKey={[Function]}
  size="middle"
/>
`;

exports[`DetailTable render renders given rows and columns 1`] = `
<ForwardRef(InternalTable)
  columns={
    Array [
      Object {
        "dataIndex": "col1",
        "filterDropdown": [Function],
        "filterIcon": [Function],
        "key": "col1",
        "onCell": [Function],
        "onFilter": [Function],
        "onHeaderCell": [Function],
        "render": [Function],
        "sorter": [Function],
        "title": "col1",
      },
      Object {
        "dataIndex": "col0",
        "filterDropdown": [Function],
        "filterIcon": [Function],
        "key": "col0",
        "onCell": [Function],
        "onFilter": [Function],
        "onHeaderCell": [Function],
        "render": [Function],
        "sorter": [Function],
        "title": "col0",
      },
    ]
  }
  dataSource={
    Array [
      Object {
        "col0": "val0",
        "col1": "val1",
      },
      Object {
        "col0": "val2",
        "col1": "val3",
      },
    ]
  }
  key="table"
  pagination={false}
  rowKey={[Function]}
  size="middle"
/>
`;
